pandas/indexes/multi.py

```diff
@@ -867,9 +865,7 @@ def from_arrays(cls, arrays, sortorder=None, names=None):
             if len(arrays[i]) != len(arrays[i - 1]):
                 raise ValueError('all arrays must be same length')
 
-        cats = [Categorical.from_array(arr, ordered=True) for arr in arrays]
-        levels = [c.categories for c in cats]
-        labels = [c.codes for c in cats]
+        labels, levels = zip(*[_factorize(arr) for arr in arrays])
```

Comment:
A quick search gives more results - there're also `stack`, `concat` and `panel_index` and probably some more. I'll try to gather them all in a separate comment.
Maybe this `_factorize` or/and a higher level `_factorize_from_iterables` should rather go to another location - `core/common.py`?

Implementation-wise, at first I was thinking about introducing a boolean parameter to `Categorical`, say `preserve_categorical_dtype` or the like. When invoked 
```
c = pd.Categorical(cat, preserve_categorical_dtype=True)
```
with a categorical array/index/series `cat`, the `c.categories` - inherited from `cat` - would be wrapped in a CategoricalIndex. (This way all changes would be contained in the Categorical constructor and no external function `_factorize` would be needed.) Would it be an acceptable solution?